### PR TITLE
lexer: implement \x and \u numeric escape sequences

### PR DESCRIPTION
**Pull request by gabrielhnf**
Implements the two `todo!()` stubs in `parse_escape` for numeric escape sequences:

- `\xNN` produces the byte with that hex value. Non-POSIX, gated on !posix_strict.
- `\uNNNN` produces the UTF-8 encoding of the codepoint.

Both follow gawk's behavior of passing the letter through literally when no hex digits follow.

Also adds tests covering single/double digit hex, uppercase digits, multi-byte unicode, and the no-digits edge case for both escapes.

### CODE REVIEW

**Comment by Alonely0** [comment]
Amazing, tysm! Let me get it reviewed this afternoon and get back to you. Do you mind fixing the rustfmt CI in the meantime?

**Comment by oech3** [comment]
Why does not pre-commit.ci `cargo fmt`? I might look at coreutils's repo.

**Comment by gabrielhnf** [comment]
> Amazing, tysm! Let me get it reviewed this afternoon and get back to you. Do you mind fixing the rustfmt CI in the meantime?

Thanks! Should be fixed now, I always forget about it

File: lexer/src/lib.rs [hunk 1]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+                            b'A'..=b'F' => (digit - b'A' + 10) as u32,
+                            _ => unreachable!(),
+                        }
+                });
+
+                let c = char::from_u32(codepoint).ok_or(LexingError::Unknown)?; // or a more specific error
+
+                let mut buf = [0u8; 4];
+                let encoded = c.encode_utf8(&mut buf);
+                out.extend_from_slice(encoded.as_bytes());
```
Comment (Alonely0) [inline lexer/src/lib.rs]: Consider switching to `extend_from_slice_copy`. Hopefully this method is better at telling LLVM to do it in-place.

File: lexer/src/lib.rs [hunk 2]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+                        + match digit {
+                            b'0'..=b'9' => digit - b'0',
+                            b'a'..=b'f' => digit - b'a' + 10,
+                            b'A'..=b'F' => digit - b'A' + 10,
+                            _ => unreachable!(),
+                        }
+                }) as char
+            }
+        }
+        'u' => {
```
Comment (Alonely0) [inline lexer/src/lib.rs]: It appears this is also gated under non-POSIX.

File: lexer/src/lib.rs [hunk 3]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+                            b'a'..=b'f' => (digit - b'a' + 10) as u32,
+                            b'A'..=b'F' => (digit - b'A' + 10) as u32,
+                            _ => unreachable!(),
+                        }
+                });
+
+                let c = char::from_u32(codepoint).ok_or(LexingError::Unknown)?; // or a more specific error
+
+                let mut buf = [0u8; 4];
+                let encoded = c.encode_utf8(&mut buf);
```
Comment (Alonely0) [inline lexer/src/lib.rs:451]: Assumes the system locale is UTF-8. This is fine for now; we can do this in a subsequent PR given its breadth. However, note that there are quite a few assumptions about UTF-8 in the code.

File: lexer/src/lib.rs [hunk 4]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+            }
+        }
+        'u' => {
+            let is_hex = |i: usize| {
+                slice
+                    .get(i)
+                    .is_some_and(|&x| (x as char).is_ascii_hexdigit())
+            };
+
+            let num_digits = (2..=5).take_while(|&i| is_hex(i)).count();
```
Comment (Alonely0) [inline lexer/src/lib.rs]: Note this takes up to 4 characters and awk takes up to 8 (related to it being locale-dependant and not tied to UTF-8). Reproducible string: `"\u00000032"`; should output `2`, as in `"\u32"`.

File: lexer/src/lib.rs [hunk 5]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+                'u'
+            } else {
+                count += num_digits;
+
+                let codepoint = slice[2..2 + num_digits].iter().fold(0u32, |acc, &digit| {
+                    acc * 16
+                        + match digit {
+                            b'0'..=b'9' => (digit - b'0') as u32,
+                            b'a'..=b'f' => (digit - b'a' + 10) as u32,
+                            b'A'..=b'F' => (digit - b'A' + 10) as u32,
```
Comment (Alonely0) [inline lexer/src/lib.rs]: nitpick: I'm not thrilled about the unreachable but it should be OK enough for LLVM to optimize away.

File: lexer/src/lib.rs [hunk 6]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+            } else {
+                count += num_digits;
+
+                slice[2..2 + num_digits].iter().fold(0u8, |acc, &digit| {
+                    acc * 16
+                        + match digit {
+                            b'0'..=b'9' => digit - b'0',
+                            b'a'..=b'f' => digit - b'a' + 10,
+                            b'A'..=b'F' => digit - b'A' + 10,
+                            _ => unreachable!(),
```
Comment (Alonely0) [inline lexer/src/lib.rs]: Ditto for the comments on the other match. Consider moving this to an utility freestanding function?

File: lexer/src/lib.rs [hunk 7]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
                 .take(count - 1)
                 .fold(0, |acc, digit| acc * 8 + digit - b'0') as char
         }
-        'x' if !posix_strict => todo!(),
-        'u' => todo!(),
+        'x' if !posix_strict => {
+            let is_hex = |i: usize| {
```
Comment (Alonely0) [inline lexer/src/lib.rs]: It would be best to move this to the top of the function to be reused by `\u`, or as a freestanding function. Do what you think fits better.

File: lexer/src/lib.rs [hunk 8]
```diff
@@ -417,8 +417,63 @@ fn parse_escape<const REGEX: bool>(
...
+                    acc * 16
+                        + match digit {
+                            b'0'..=b'9' => (digit - b'0') as u32,
+                            b'a'..=b'f' => (digit - b'a' + 10) as u32,
+                            b'A'..=b'F' => (digit - b'A' + 10) as u32,
+                            _ => unreachable!(),
+                        }
+                });
+
+                let c = char::from_u32(codepoint).ok_or(LexingError::Unknown)?; // or a more specific error
```
Comment (Alonely0) [inline lexer/src/lib.rs]: Note: this also makes UTF-8 assumptions; and `\u` in gawk never errors, it inserts the locale's replacement character (`U+FFFD` in UTF-8, fwiw) or `?` for lack thereof.

**Review by Alonely0 — CHANGES_REQUESTED** [review]
Overall, it's pretty good. However, there are a few things that need to be changed. Please, take a look at the [docs](https://www.gnu.org/software/gawk/manual/html_node/Escape-Sequences.html). As for the UTF-8 specific bits; keep them (we can do it in a subsequent PR), but add `FIXME` comments about it in the relevant places so we know about this. Please, also add tests that test the behavior of both flags for POSIX conformance.

File: lexer/src/lib.rs (hunk 8 above)
Comment (gabrielhnf) [reply lexer/src/lib.rs]: Since we're assuming UTF-8 for now, I took the liberty of going with `U+FFFD` on error. Let me know if you prefer something different

File: lexer/src/lib.rs (hunk 5 above)
Comment (gabrielhnf) [reply lexer/src/lib.rs]: I'm not sure what change you have in mind for the `unreachable!`, so I left it as is for now, happy to update if you have a preference

**Review by Alonely0 — APPROVED** [review]

**Comment by Alonely0** [comment]
That looks great, tysm!
